slkspec/core.py

```diff
@@ -151,6 +151,18 @@ def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
             for out_file in map(Path, inp_files):
                 (output_dir / out_file.name).chmod(self.file_permissions)
 
+    def mkdirs(self, path):
+        rp = os.path.realpath(path)
+        if os.access(rp, os.F_OK):
+            return
+        components = Path(rp).parts[1:]
+        for i in range (len(components)):
+            subpath = Path("/", *components[:i+1])
+            if not os.access(subpath, os.F_OK):
```

Comment:
Shouldn't this rather be `os.W_OK`? Or even better rewritten to a try/except statement? Currently, a folder might exists, but `os.mkdir` still fail because the user doesn't have write permissions.

Reply:
No, we create a subdir, if it does not exist. 
We don't try to create /scratch/ or so, and we don't have write permissions there anyways.

We could again add checks for write permissions starting at the point where we need to start creating (parts of) the path.

Basically the issue here is finding out where in /path/to/data/in/a/subdir/ we have to start creating and adjusting permissions, and the easiest way, I could come up with for doing that was to go bit by bit, check if the subdir is there and create and chmod if necessary.

Reply:
The issue of us needing weird subdirectory permissions actually remains independent of the pyslk version...
Group writable sticky stuff is not totally standard, but very helpful for this game with multiple users.
So we might actually have to start putting more rather than less thought into this.

Reply:
Well, I would like to keep `slkspec` as general as possible and also useful for other slk tape archives. In the best case only small changes need to be made to use `slkspec` with other tape archive that just offer any way to retrieve files. Introducing handling of the `sticky bit` is very specific and I'd prefer to implement this into the DKRZ-SLK version and not in the last step. Retrieving a file with `slkspec` and `slk retrieve` should result in the same result.

Reply:
Could you maybe raise this issue upstream, so we could just forward some specific keywords to `pyslk`?

Reply:
CC @neumannd

Reply:
@observingClouds Is this a feature request for `pyslk` or `slk`? This is no request to add the sticky bit to StrongLink but just for `slk`/`pyslk` to set it for retrieved folders?

You mean a sticky bit as defined as follows (source: https://unix.stackexchange.com/a/79401/182655)?

> The restricted deletion flag or sticky bit is a single bit, whose interpretation depends on the file type. For directories, it
prevents unprivileged users from removing or renaming a file in the directory unless they own the file or the directory; this is called the restricted deletion flag for the directory, and is commonly found on world-writable directories like /tmp.

Reply:
@florianziemen seems to be interested in the sticky bit and I think it would be nice if all retrieval methods ( `slk retrieve`, `pyslk.core.retrieve` and `slkspec` behave in the same way. Because all these tools build on top of each other, it would be great if the implementation would be as much down the stack as possible.

Reply:
`cp` does not support setting the sticky bit while files are copied. There the user needs to set permissions afterwards. `rsync` offers arguments `--chmod` and `--chown`. We could request something similar. What is the advantage of adding this argument to `slk retrieve` compared to just running `chmod` after `slk retrieve?`

Reply:
Moin!

So, the really important part ist the gid bit for folders. (2000 of 3775 or so), because it means that subfolders are created with the same user group as the original folder. So, if I download a file into CACHE/arch/run/subdir, subdir belongs to the "intake group", otherwise it becomes part of my group (DKRZ ANW) and no "normal" users can write there (similarly for MPI-ATM vs MPI-OCE / ...).
The sticky bit is nice and seems fitting for the purpose but is less important to me.

So, basically, what we need to do is figure out which folders slk is going to create, and make those gid-bit-y (or change their groups).

Reply:
What if this is applied on the top folder:

```
chgrp -R ${group} ${folder}
chmod g+rwxs -R ${folder}
```

This should set `$group` permanently for all newly created sub-folders of `$folder`. At least that's what I did with my folder into which I retrieve data for users:

```
chgrp -R ka1209 ${folder}
chmod g+rwxs -R ${folder}
umask 002 ${folder}
setfacl -Rdm g::rwx  ${folder}
```

All users have read permissions and a specific group has write permissions.

I just checked but it seems that the sticky bit of a folder cannot be inherited by its sub-folder.

Reply:
yep, so the gid bit ( 2000 ) is passed down. Ok...
Whether we want the suid bit for folders (definitely not for files) is a bit unclear to me.
If we have it, people can't remove other people's files. Seems reasonable (even just to prevent accidental deletions), but also means, I can't get rid of somebody else's broken file if something got stuck mid-way.
We always have the option of calling for divine intervention, but that should not be the default use case.

The other thing we need is the directories to stay group-writable. For me, the umask will generally prevent this, so we need something that takes care of creating/chowning the directories that slk creates upon downloading (assuming slk respects my umask).
Either (1) we are faster and simply create them before slk starts, or (2) we have to _clean up_ after it. I think, I would prefer option (1) as it is more compatible with retrievals by different users and crashed jobs.